plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/BaseJdbcCastPushdownTest.java

```diff
@@ -71,24 +71,55 @@ public void testJoinPushdownWithCast()
     @Test
     public void testCastFails()
     {
-        for (CastTestCase testCase : failCast()) {
-            assertThatThrownBy(() -> getQueryRunner().execute("SELECT CAST(%s AS %s) FROM %s".formatted(testCase.sourceColumn(), testCase.castType(), leftTable())))
-                    .hasMessageMatching("(.*)Cannot cast (.*) to (.*)");
+        Session withoutPushdown = Session.builder(getSession())
+                .setSystemProperty("allow_pushdown_into_connectors", "false")
+                .build();
+
+        for (FailCastTestCase testCase : failCast()) {
+            if (testCase.pushdownErrorMessage().isPresent()) {
+                assertThat(query("SELECT CAST(%s AS %s) FROM %s".formatted(testCase.sourceColumn(), testCase.castType(), leftTable())))
+                        .failure()
+                        .hasMessageMatching(testCase.pushdownErrorMessage().get());
+                assertThat(query(withoutPushdown, "SELECT CAST(%s AS %s) FROM %s".formatted(testCase.sourceColumn(), testCase.castType(), leftTable())))
+                        .failure()
+                        .hasMessageMatching(testCase.errorMessage());
+            }
+            else {
+                assertThat(query("SELECT CAST(%s AS %s) FROM %s".formatted(testCase.sourceColumn(), testCase.castType(), leftTable())))
+                        .failure()
+                        .hasMessageMatching(testCase.errorMessage());
+            }
         }
     }
 
     public record CastTestCase(String sourceColumn, String castType, Optional<String> targetColumn)
     {
-        public CastTestCase(String sourceColumn, String castType)
+        public CastTestCase
         {
-            this(sourceColumn, castType, Optional.empty());
+            requireNonNull(sourceColumn, "sourceColumn is null");
+            requireNonNull(castType, "castType is null");
+            requireNonNull(targetColumn, "targetColumn is null");
         }
+    }
 
-        public CastTestCase
+    public record FailCastTestCase(String sourceColumn, String castType, String errorMessage, Optional<String> pushdownErrorMessage)
```

Comment:
Why not to use different factory methods for CastTestCase ?

Reply:
I think its better to have two separate records for success and failed cases. As `targetColumn` is not necessary in failed case and `errorMessage`, `pushdownErrorMessage` is not required for success case.

Introducing factory method

```
public record CastTestCase(String sourceColumn, String castType, Optional<String> targetColumn, Optional<String> errorMessage, Optional<String> pushdownErrorMessage)
{
  public CastTestCase
  {
      requireNonNull(sourceColumn, "sourceColumn is null");
      requireNonNull(castType, "castType is null");
      requireNonNull(targetColumn, "targetColumn is null");
      requireNonNull(targetColumn, "errorMessage is null");
      requireNonNull(targetColumn, "pushdownErrorMessage is null");
  }

  public CastTestCase castTestCase(String sourceColumn, String castType, String targetColumn)
  {
      return new CastTestCase(sourceColumn, castType, Optional.of(targetColumn), Optional.empty(), Optional.empty());
  }

  public CastTestCase failingCastTestCase(String sourceColumn, String castType)
  {
      return new CastTestCase(sourceColumn, castType, Optional.empty(), Optional.of("(.*)Cannot cast (.*) to (.*)"), Optional.empty());
  }

  public CastTestCase failingCastTestCase(String sourceColumn, String castType, String errorMessage)
  {
      return new CastTestCase(sourceColumn, castType, Optional.empty(), Optional.of(errorMessage), Optional.empty());
  }

  public CastTestCase failingCastTestCase(String sourceColumn, String castType, String errorMessage, String pushdownErrorMessage)
  {
      return new CastTestCase(sourceColumn, castType, Optional.empty(), Optional.of(errorMessage), Optional.of(pushdownErrorMessage));
  }
}
```

By having 2 separate records.
```
public record CastTestCase(String sourceColumn, String castType, String targetColumn)
{
    public CastTestCase
    {
        requireNonNull(sourceColumn, "sourceColumn is null");
        requireNonNull(castType, "castType is null");
        requireNonNull(targetColumn, "targetColumn is null");
    }
}

public record FailCastTestCase(String sourceColumn, String castType, String errorMessage, Optional<String> pushdownErrorMessage)
{
    public FailCastTestCase(String sourceColumn, String castType)
    {
        this(sourceColumn, castType, "(.*)Cannot cast (.*) to (.*)");
    }

    public FailCastTestCase(String sourceColumn, String castType, String errorMessage)
    {
        this(sourceColumn, castType, errorMessage, Optional.empty());
    }

    public FailCastTestCase
    {
        requireNonNull(sourceColumn, "sourceColumn is null");
        requireNonNull(castType, "castType is null");
        requireNonNull(errorMessage, "errorMessage is null");
        requireNonNull(pushdownErrorMessage, "pushdownErrorMessage is null");
    }
}
```